framework/src/wifi_manager/slsi_wifi_utils_lwnl80211.c

```diff
@@ -0,0 +1,483 @@
+
+#define WU_CLOSE(fd)                            \
+	do {                                        \
+		close(fd);                              \
+	} while (0)
+
+static wifi_utils_cb_s g_cbk = {NULL, NULL, NULL, NULL, NULL};
+
+struct _wifi_utils_s g_hnd = {"", 0, 0};
+
+static void free_scan_data(wifi_utils_scan_list_s *scan_list)
+{
+	wifi_utils_scan_list_s *cur = scan_list;
+	wifi_utils_scan_list_s *prev = NULL;
+	while (cur) {
+		prev = cur;
+		cur = cur->next;
+		free(prev);
+	}
+	scan_list = NULL;
+}
+
+static wifi_utils_result_e receive_scan_data(mqd_t mq, wifi_utils_scan_list_s *scan_list)
+{
+	int ret;
+	int prio;
+	int nbytes;
+	int cnt = 0;
+	int msglen = sizeof(lwnl80211_cb_data);
+	wifi_utils_scan_list_s *prev = scan_list;
+
+	while (true) {
+		lwnl80211_cb_data msg;
+		wifi_utils_scan_list_s *cur = NULL;
+		nbytes = mq_receive(mq, (char *)&msg, msglen, &prio);
+		if (nbytes < 0 || nbytes != msglen) {
+			WU_LOG("Failed to receive (nbytes=%d, msglen=%d)\n", nbytes, msglen);
+			WU_ERR;
+			ret = WIFI_UTILS_FAIL;
+			break;
+		}
+
+		if (msg.status != LWNL80211_SCAN_DONE) {
+			WU_ERR;
+			ret =  WIFI_UTILS_FAIL;
+			break;
+		}
+
+		cur = (wifi_utils_scan_list_s *)malloc(sizeof(wifi_utils_scan_list_s));
+		if (!cur) {
+			free_scan_data(scan_list);
+			ret = WIFI_UTILS_FAIL;
+			break;
+		}
+		cur->next = NULL;
+		memcpy(&(cur->ap_info), &(msg.u.ap_info), sizeof(wifi_utils_ap_scan_info_s));
+		cnt++;
+
+		prev->next = cur;
+		prev = cur;
+
+		if (msg.md == 0 || cnt >= LWNL80211_MQUEUE_MAX_DATA_NUM) {
+			WU_LOG("End of scanning (%d data)\n", cnt);
+			ret = WIFI_UTILS_SUCCESS;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+void *wifi_utils_callback_handler(void *arg)
+{
+	WU_ENTER;
+
+	int prio;
+	int nbytes;
+	lwnl80211_cb_data msg;
+	int msglen = sizeof(lwnl80211_cb_data);
+
+	while (true) {
+		if (g_hnd.mq == NULL) {
+			break;
+		}
+
+		nbytes = mq_receive(g_hnd.mq, (char *)&msg, msglen, &prio);
+		if (nbytes < 0 || nbytes != msglen) {
+			WU_LOG("Failed to receive (nbytes=%d, msglen=%d)\n", nbytes, msglen);
+			WU_ERR;
+			break;
+		}
+
+		switch (msg.status) {
+		case LWNL80211_STA_CONNECTED:
+			g_cbk.sta_connected(WIFI_UTILS_SUCCESS, NULL);
+			break;
+		case LWNL80211_STA_CONNECT_FAILED:
+			g_cbk.sta_connected(WIFI_UTILS_FAIL, NULL);
+			break;
+		case LWNL80211_STA_DISCONNECTED:
+			g_cbk.sta_disconnected(NULL);
+			break;
+		case LWNL80211_SOFTAP_STA_JOINED:
+			g_cbk.softap_sta_joined(NULL);
+			break;
+		case LWNL80211_SOFTAP_STA_LEFT:
+			g_cbk.softap_sta_left(NULL);
+			break;
+		case LWNL80211_SCAN_FAILED:
+			g_cbk.scan_done(WIFI_UTILS_FAIL, NULL, NULL);
+			break;
+		case LWNL80211_SCAN_DONE:
+		{
+			wifi_utils_scan_list_s *scan_list;
+			scan_list = (wifi_utils_scan_list_s *)malloc(sizeof(wifi_utils_scan_list_s));
+			scan_list->next = NULL;
+			if (!scan_list) {
+				free_scan_data(scan_list);
+				WU_ERR;
+				break;
+			}
+			memcpy(&(scan_list->ap_info),&(msg.u.ap_info), sizeof(wifi_utils_ap_scan_info_s));
```

Comment:
Coding rules: Please provide space after coma.

Reply:
Applied. Thanks for the comment.